pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -465,18 +476,33 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 
 // runNextSchedulableTask gets the next schedulable Tasks from the dag based on the current
 // pipeline run state, and starts them
-func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
+// after all DAG tasks are done, its responsible for scheduling final tasks and start executing them
+func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, dfinally *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
 
 	logger := logging.FromContext(ctx)
 	recorder := controller.GetEventRecorder(ctx)
 
-	candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulPipelineTaskNames()...)
-	if err != nil {
-		logger.Errorf("Error getting potential next tasks for valid pipelinerun %s: %v", pr.Name, err)
-		return nil
+	var nextRprts []*resources.ResolvedPipelineRunTask
+
+	// when pipeline run is stopping, do not schedule any new task and only
+	// wait for all running tasks to complete and report their status
+	if !pipelineState.IsStopping(d) {
+		// candidateTasks is derived based on successfully finished tasks and/or skipped tasks
+		// candidateTasks is initialized to DAG root nodes to start pipeline execution
+		candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulOrSkippedDAGTasks(d)...)
+		if err != nil {
+			logger.Errorf("Error getting potential next tasks for valid pipelinerun %s: %v", pr.Name, err)
+			return nil
+		}
+		// nextRprts holds a list of pipeline tasks which should be executed next
+		nextRprts = pipelineState.GetNextTasks(candidateTasks)
 	}
 
-	nextRprts := pipelineState.GetNextTasks(candidateTasks)
+	// append a list of final tasks which needs to be scheduled to nextRprts
+	// final tasks are only scheduled after all DAG tasks have finished successfully
+	// or a DAG task might have failed and no other parallel tasks are running
+	nextRprts = append(nextRprts, pipelineState.GetFinalTasks(d, dfinally)...)
```

Comment:
I thought it was going to be

```
if ! isStopping {
  Get candidates
  Get next tasks
  Get final tasks and append
  Loop over next tasks
}
```

This works too, hopefully the refactor on top will make things clearer.

Reply:
tried without success 😢 Changing it to `if ! isStopping { ... }`, breaks one of the [e2e](https://github.com/tektoncd/pipeline/pull/2661/files#diff-a80234191e535ce50a353f68f8fbb21dR34) tests, specifically pipeline involving skipped tasks i.e. conditional tasks with failing condition.